README.md
# Introduction

This Koha plugin enables a library to accept online payments from patrons using the WPM payments platform.

Plugins are installed by uploading KPZ ( Koha Plugin Zip ) packages. A KPZ file is just a zip file containing the perl files, template files, and any other files necessary to make the plugin work. Learn more about the Koha Plugin System in the [Koha 3.22 Manual](http://manual.koha-community.org/3.22/en/pluginsystem.html).

# Downloading

From the [release page](https://github.com/ptfs-europe/koha-plugin-wpmpayments/releases) you can download the relevant *.kpz file

# Installing

To set up the Koha plugin system you must first make some changes to your install.

* Change `<enable_plugins>0<enable_plugins>` to `<enable_plugins>1</enable_plugins>` in your koha-conf.xml file
* Confirm that the path to `<pluginsdir>` exists, is correct, and is writable by the web server
* Add the pluginsdir to your apache PERL5LIB paths and koha-plack startup scripts PERL5LIB
* Restart your webserver

Once set up is complete you will need to alter your UseKohaPlugins system preference. On the Tools page you will see the Tools Plugins and on the Reports page you will see the Reports Plugins.

# Setup

You will need to add to the apache config for your site:
```
   Alias /plugin/ "/var/lib/koha/kohadev/plugins/"
   # The stanza below is needed for Apache 2.4+
   <Directory /var/lib/koha/kohadev/plugins/>
         Options Indexes FollowSymLinks
         AllowOverride None
         Require all granted
         Options +ExecCGI
         AddHandler cgi-script .pl
    </Directory>
```

You may also need to add to PERL5LIB


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [v00.00.12] - 2023-12-04
- Handle "pretty" KohaAdminEmailAddress

## [v00.00.11] - 2023-11-08
- Style improvements for 22.11
- Pretty printing of XML in logs

## [v00.00.10] - 2022-07-20

### Fixed
- Ensure all used methods are imported - 21.11 fix

## Enhanced
- Update templates for 21.11 display improvements

## [v00.00.09] - 2021-05-27

### Fixed
- Add missing closing brace to logic statement that prevented return of callback.pl.. doh!

## [v00.00.08] - 2021-05-26

### Fixed
- Add missing semi-colon that prevented compile.. doh!

## [v00.00.07] - 2021-05-26

### Fixed
- Correct version comparision to use internal perl version objects for dotted decimal version string comparisons
- Updated custom field regex to be more lenient with whitespace and plurals
- Map debit_type_codes for system types to allow backward compatability for WPM
- Set a userenv in callback.pl so that we do not fail on unfixed versions of Koha::Account::pay
- Update for change in return signature of Koha::Account::pay in 20.05.00 release
- Tested against Koha 20.11.06

## [v00.00.06] - 2020-08-20

### Fixed
- Add support for 19.11.x (accounttype -> debit_type_code)

## [v00.00.05] - 2019-09-06

### Added
- Configurable transaction level customfield definitions including basic template support
- Added the `interface` argument to the parameters passed in the call to Koha::Account::pay

## [v00.00.04] - 2019-08-09

### Fixed
- Remove $cgi->param called in list context warnings from configuration code
- Fixed typo in configurable VAT field keys

## [v00.00.03] - 2019-07-05

### Added
- Configurable VAT fields
- Configurable CustomField1

## [v00.00.02] - 2019-06-17

### Fixed
- Respect renewal rules.

## [v00.00.01] - 2019-06-14

Initial release

### Added
- Migrated logic from custom koha code into a plugin.
- Configuration screens
- Plugin methods for all basic actions




gulpfile.js
const gulp = require('gulp');
const release = require('gulp-github-release');
const fs = require('fs');
const run = require('gulp-run');
const dateTime = require('node-datetime');

const dt = dateTime.create();
const today = dt.format('Y-m-d');

const package_json = JSON.parse(fs.readFileSync('./package.json'));
const release_filename = `${package_json.name}-v${package_json.version}.kpz`;

const pm_file = 'WPMPayments.pm';
const pm_file_path = 'Koha/Plugin/Com/PTFSEurope/';
const pm_file_path_full = pm_file_path + pm_file;
const pm_file_path_dist = 'dist/' + pm_file_path;
const pm_file_path_full_dist = pm_file_path_dist + pm_file;

console.log(release_filename);
console.log(pm_file_path_full_dist);

gulp.task('build', () => {
    run(`
        mkdir dist ;
        cp -r Koha dist/. ;
        sed -i -e "s/{VERSION}/${package_json.version}/g" ${pm_file_path_full_dist} ;
        sed -i -e "s/1900-01-01/${today}/g" ${pm_file_path_full_dist} ;
        cd dist ;
        zip -r ../${release_filename} ./Koha ;
        cd .. ;
        rm -rf dist ;
    `).exec();

});

gulp.task('release', () => {
    gulp.src(release_filename)
        .pipe(release({
            manifest: require('./package.json') // package.json from which default values will be extracted if they're missing
        }));
});


release_kpz.sh
#!/bin/bash

echo "Creating and releasing .kpz file"

FILE=./koha-plugin-wpm-payments.kpz
if test -f "$FILE"; then
    echo "Plugin .kpz file already exists, previous version deleted"
    rm -r koha-plugin-wpm-payments.kpz
fi
echo "Creating new .kpz file"
zip -r koha-plugin-wpm-payments.kpz Koha/

NEW_VERSION=$(node checkVersionNumber.js)
NEW_VERSION_NUMBER=${NEW_VERSION:1}
PREVIOUS_VERSION=$(git log -1 --pretty=oneline | grep -E -o 'v.{0,4}')
PREVIOUS_VERSION_NUMBER=${PREVIOUS_VERSION:-1}
echo "New version: ${NEW_VERSION_NUMBER}"
echo "Previous version: ${PREVIOUS_VERSION_NUMBER}"

if [ $NEW_VERSION_NUMBER != $PREVIOUS_VERSION_NUMBER ]; then
    echo "Version has been updated - starting upload"
    git add .
    git commit -m "$NEW_VERSION"
    git push
    echo "Plugin has been pushed to Github and a release is being generated"
else
    echo "WARNING: The Plugin version needs to be updated - please check the .pm file and update the version"
fi


Koha/Plugin/Com/PTFSEurope/WPMPayments/callback.pl
#!/usr/bin/perl

# Copyright 2015 PTFS Europe
#
# This file is part of Koha.
#
# Koha is free software; you can redistribute it and/or modify it under the
# terms of the GNU General Public License as published by the Free Software
# Foundation; either version 3 of the License, or (at your option) any later
# version.
#
# Koha is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
# A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with Koha; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

use Modern::Perl;

use CGI qw( -utf8 );

use C4::Context;
use C4::Circulation;
use Koha::Account;
use Koha::Account::Lines;
use Koha::Patrons;
use Koha::Plugin::Com::PTFSEurope::WPMPayments;

use XML::LibXML;
use Digest::MD5 qw(md5_hex);

my $debug = 0;

my $paymentHandler = Koha::Plugin::Com::PTFSEurope::WPMPayments->new;

# Parse XML
binmode STDIN, ':encoding(UTF-8)';
my $xml_string;
while (<STDIN>) {
    $xml_string .= $_;
}
my $xml;
eval { $xml = XML::LibXML->load_xml( string => $xml_string ); };
warn "error: " . $@ if $@;

my $borrowernumber = $xml->findvalue('/wpmpaymentrequest/customerid');
my $transaction_id = $xml->findvalue('/wpmpaymentrequest/transactionreference');
my $success        = $xml->findvalue('/wpmpaymentrequest/transaction/success');

my $borrower = Koha::Patrons->find($borrowernumber);

# Set the userenv
C4::Context->_new_userenv( 'PLUGIN_' . time() );
C4::Context->set_userenv(
    $borrower->borrowernumber, $borrower->userid,
    $borrower->cardnumber,     $borrower->firstname,
    $borrower->surname,        $borrower->branchcode,
    $borrower->flags,          undef,
    undef,                     undef,
    undef,
);

if ( $success eq '1' ) {

    $debug and warn "Recieved 'success' from WPM";

    # Extract accountlines to pay
    my @accountline_ids = ();
    my $payments_nodes  = $xml->findnodes('/wpmpaymentrequest/payments');
    for my $payments_node ( $payments_nodes->get_nodelist ) {
        my $payment_nodes = $payments_node->findnodes('./payment');
        for my $payment_node ( $payment_nodes->get_nodelist ) {
            if ( $payment_node->findvalue('./@paid') ) {
                my $accountline = $payment_node->findvalue('./@payid');
                push @accountline_ids, $accountline;
            }
        }
    }

    my $totalpaid = $xml->findvalue('/wpmpaymentrequest/transaction/totalpaid');

    # Make Payment
    my $lines = Koha::Account::Lines->search(
        { accountlines_id => { 'in' => \@accountline_ids } } )->as_list;
    my $account        = Koha::Account->new( { patron_id => $borrowernumber } );
    my $accountline_id = $account->pay(
        {
            amount     => $totalpaid,
            note       => 'WPM Payment',
            library_id => $borrower->branchcode,
            interface  => 'opac',
            lines => $lines,    # Arrayref of Koha::Account::Line objects to pay
            #payment_type => $payment_type,  # accounttype code
        }
    );
    $debug
      and warn "Payment of $totalpaid made against "
      . join( ', ', @accountline_ids );

    # Return signature of ->pay changed with version 20.05.00
    if ( $paymentHandler->_version_check('20.05.00') ) {
        $accountline_id = $accountline_id->{payment_id};
    }

    # Link payment to wpm_transactions
    my $dbh   = C4::Context->dbh;
    my $table = $paymentHandler->get_qualified_table_name('wpm_transactions');
    my $sth   = $dbh->prepare(
        "UPDATE $table SET accountline_id = ? WHERE transaction_id = ?");
    $sth->execute( $accountline_id, $transaction_id );
    $debug and warn "Update the original transaction";
    $debug
      and warn
"UPDATE $table SET accountline_id = $accountline_id WHERE transaction_id = $transaction_id;";

    # Renew any items as required
    unless ( $paymentHandler->_version_check('20.05.00') ) {
        for my $line ( @{$lines} ) {
            my $item = Koha::Items->find( { itemnumber => $line->itemnumber } );

            # Renew if required
            if ( $paymentHandler->_version_check('19.11.00') ) {
                if (   $line->debit_type_code eq "OVERDUE"
                    && $line->status ne "UNRETURNED" )
                {
                    if (
                        C4::Circulation::CheckIfIssuedToPatron(
                            $line->borrowernumber, $item->biblionumber
                        )
                      )
                    {
                        my ( $renew_ok, $error ) =
                          C4::Circulation::CanBookBeRenewed(
                            $line->borrowernumber, $line->itemnumber, 0 );
                        if ($renew_ok) {
                            C4::Circulation::AddRenewal( $line->borrowernumber,
                                $line->itemnumber );
                        }
                    }
                }
            }
            else {
                if ( defined( $line->accounttype )
                    && $line->accounttype eq "FU" )
                {
                    if (
                        C4::Circulation::CheckIfIssuedToPatron(
                            $line->borrowernumber, $item->biblionumber
                        )
                      )
                    {
                        my ( $can, $error ) =
                          C4::Circulation::CanBookBeRenewed(
                            $line->borrowernumber, $line->itemnumber, 0 );
                        if ($can) {

                            # Fix paid for fine before renewal to prevent
                            # call to _CalculateAndUpdateFine if
                            # CalculateFinesOnReturn is set.
                            C4::Circulation::_FixOverduesOnReturn(
                                $line->borrowernumber, $line->itemnumber );

                            # Renew the item
                            my $datedue =
                              C4::Circulation::AddRenewal(
                                $line->borrowernumber, $line->itemnumber );

                            $debug
                              and warn
                              "Renewal of $line->itemnumber successful";
                        }
                        else {
                            $debug
                              and warn "Renewal of $line->itemnumber blocked";
                        }
                    }
                }
            }
        }
    }

    # Respond with OK
    my $response = new CGI;
    my $reply    = XML::LibXML::Document->new( '1.0', 'utf-8' );
    my $root     = $reply->createElement("wpmmessagevalidation");
    my $md5      = $xml->findvalue('/wpmpaymentrequest/@msgid');
    $root->setAttribute( 'msgid' => "$md5" );

    my $validation = $reply->createElement('validation');
    $validation->appendTextNode("1");
    $root->appendChild($validation);

    my $validationmessage = $reply->createElement('validationmessage');
    my $success           = XML::LibXML::CDATASection->new("Success");
    $validationmessage->appendChild($success);
    $root->appendChild($validationmessage);

    $reply->setDocumentElement($root);

    print CGI->header('text/xml');
    print $reply->toString();
}
else {
    # Update transaction status
    #
    # Respond OK

    my $reply = XML::LibXML::Document->new( '1.0', 'utf-8' );
    my $root  = $reply->createElement("wpmmessagevalidation");
    my $md5   = $xml->findvalue('/wpmpaymentrequest/@msgid');
    $root->setAttribute( 'msgid' => "$md5" );

    my $validation = $reply->createElement('validation');
    $validation->appendTextNode("1");
    $root->appendChild($validation);

    my $validationmessage = $reply->createElement('validationmessage');
    my $success           = XML::LibXML::CDATASection->new("Success");
    $validationmessage->appendChild($success);
    $root->appendChild($validationmessage);

    $reply->setDocumentElement($root);

    print CGI->header('text/xml');
    print $reply->toString();
}


Koha/Plugin/Com/PTFSEurope/WPMPayments.pm
use utf8;

package Koha::Plugin::Com::PTFSEurope::WPMPayments;

use Modern::Perl;

use base qw(Koha::Plugins::Base);
use version 0.77;

use C4::Context;
use C4::Auth qw( get_template_and_user );
use Koha::Account;
use Koha::Account::Lines;
use Koha::Patrons;

use Email::Address;
use XML::LibXML;
use DateTime;
use Digest::MD5 qw(md5_hex);
use HTML::Entities;

## Here we set our plugin version
our $VERSION = "00.00.12";
our $debug   = 0;

## Here is our metadata, some keys are required, some are optional
our $metadata = {
    name            => 'WPM Online Payments Plugin',
    author          => 'Martin Renvoize',
    date_authored   => '2018-06-13',
    date_updated    => "2023-07-13",
    minimum_version => '17.11.00.000',
    maximum_version => '22.11.24.000',
    version         => $VERSION,
    description     => 'This plugin implements online payments using '
      . 'WPM Educations payments platform.',
};

## Map Koha system debit types to WPM fee_Type codes
# WPM refuse to update their implimentation to reflect changes to Koha
# so we have this mapping for compatability.
our %system_type_map = (
    'OVERDUE' => 'F',
    'LOST'    => 'L'
);

sub new {
    my ( $class, $args ) = @_;

    ## We need to add our metadata here so our base class can access it
    $args->{'metadata'} = $metadata;
    $args->{'metadata'}->{'class'} = $class;

    ## Here, we call the 'new' method for our base class
    ## This runs some additional magic and checking
    ## and returns our actual $self
    my $self = $class->SUPER::new($args);

    return $self;
}

sub _version_check {
    my ( $self, $minversion ) = @_;

    my $kohaversion = Koha::version();
    return ( version->parse($kohaversion) > version->parse($minversion) );
}

sub opac_online_payment {
    my ( $self, $args ) = @_;

    return $self->retrieve_data('enable_opac_payments') eq 'Yes';
}

=head2 opac_online_payment_begin

  Initiate online payment process

=cut

sub opac_online_payment_begin {
    my ( $self, $args ) = @_;
    $debug and warn "Inside opac_online_payment_begin for: " . caller . "\n";

    my $cgi    = $self->{'cgi'};
    my $schema = Koha::Database->new()->schema();

    my ( $template, $borrowernumber ) = get_template_and_user(
        {
            template_name   => $self->mbf_path('opac_online_payment_begin.tt'),
            query           => $cgi,
            type            => 'opac',
            authnotrequired => 0,
            is_plugin       => 1,
        }
    );

    # Get the borrower
    my $borrower_result = Koha::Patrons->find($borrowernumber);

    # Create a transaction
    my $dbh   = C4::Context->dbh;
    my $table = $self->get_qualified_table_name('wpm_transactions');
    my $sth = $dbh->prepare("INSERT INTO $table (`accountline_id`) VALUES (?)");
    $sth->execute(undef);

    my $transaction_id =
      $dbh->last_insert_id( undef, undef, qw(wpm_transactions transaction_id) );

    # Construct redirect URI
    my $redirect_url = URI->new( C4::Context->preference('OPACBaseURL')
          . "/cgi-bin/koha/opac-account-pay-return.pl" );
    $redirect_url->query_form(
        {
            payment_method => scalar $cgi->param('payment_method'),
            transaction_id => $transaction_id
        }
    );

    # Construct callback URI
    my $callback_url =
      URI->new( C4::Context->preference('OPACBaseURL')
          . $self->get_plugin_http_path()
          . "/callback.pl" );

    # Construct cancel URI
    my $cancel_url = URI->new( C4::Context->preference('OPACBaseURL')
          . "/cgi-bin/koha/opac-account.pl" );

    # Construct custom fields
    my %customfields;
    for my $i ( 1 .. 10 ) {
        $customfields{$i} = $self->retrieve_data("customfield$i");
        if ( $customfields{$i} =~ m/(.*)\[% *borrowers?\.(.*?) *%\](.*)/ ) {
            my ( $pre, $variable, $post ) = ( $1, $2, $3 );
            $customfields{$i} = $pre . $borrower_result->$variable . $post;
        }
    }

    # Construct XML POST
    my $xml  = XML::LibXML::Document->new( '1.0', 'utf-8' );
    my $root = $xml->createElement('wpmpaymentrequest');

    # Email handling
    my @addrs = Email::Address->parse(C4::Context->preference('KohaAdminEmailAddress'));
    my $from_address = $addrs[0]->address;

    my @fields = (
        {
            name  => 'clientid',
            value =>
              { value => $self->retrieve_data('WPMClientID'), cdata => 0 }
        },
        {
            name  => 'requesttype',
            value => { value => 1, cdata => 0 }
        },
        {
            name  => 'pathwayid',
            value =>
              { value => $self->retrieve_data('WPMPathwayID'), cdata => 0 }
        },
        {
            name  => 'departmentid',
            value => {
                value => $self->retrieve_data('WPMDepartmentID'),
                cdata => 0
            }
        },
        { name => 'staffid', value => { value => undef, cdata => 0 } },
        {
            name  => 'customerid',
            value => { value => $borrowernumber, cdata => 0 }
        },
        {
            name  => 'title',
            value => { value => $borrower_result->title, cdata => 1 }
        },
        {
            name  => 'firstname',
            value => { value => $borrower_result->firstname, cdata => 1 }
        },
        {
            name  => 'middlename',
            value => { value => $borrower_result->othernames, cdata => 1 }
        },
        {
            name  => 'lastname',
            value => { value => $borrower_result->surname, cdata => 1 }
        },
        {
            name  => 'emailfrom',
            value => {
                value => $from_address,
                cdata => 1
            }
        },
        {
            name  => 'toemail',
            value => { value => $borrower_result->email, cdata => 1 }
        },
        { name => 'ccemail', value => { value => undef, cdata => 1 } },
        {
            name  => 'transactionreference',
            value => { value => $transaction_id, cdata => 1 }
        },
        {
            name  => 'redirecturl',
            value => { value => $redirect_url, cdata => 1 }
        },
        {
            name  => 'callbackurl',
            value => { value => $callback_url, cdata => 1 }
        },
        {
            name  => 'cancelurl',
            value => { value => $cancel_url, cdata => 1 }
        },
        {
            name  => 'billaddress1',
            value => { value => $borrower_result->streetnumber, cdata => 1 }
        },
        {
            name  => 'billaddress2',
            value => { value => $borrower_result->address, cdata => 1 }
        },
        {
            name  => 'billaddress3',
            value => { value => $borrower_result->address2, cdata => 1 }
        },
        {
            name  => 'billtown',
            value => { value => $borrower_result->city, cdata => 1 }
        },
        {
            name  => 'billcounty',
            value => { value => $borrower_result->state, cdata => 1 }
        },
        {
            name  => 'billpostcode',
            value => { value => $borrower_result->zipcode, cdata => 1 }
        },
        {
            name  => 'billcountry',
            value => { value => $borrower_result->country, cdata => 1 }
        },
        {
            name  => 'customfield1',
            value => { value => $customfields{1}, cdata => 1 }
        },
        {
            name  => 'customfield2',
            value => { value => $customfields{2}, cdata => 1 }
        },
        {
            name  => 'customfield3',
            value => { value => $customfields{3}, cdata => 1 }
        },
        {
            name  => 'customfield4',
            value => { value => $customfields{4}, cdata => 1 }
        },
        {
            name  => 'customfield5',
            value => { value => $customfields{5}, cdata => 1 }
        },
        {
            name  => 'customfield6',
            value => { value => $customfields{6}, cdata => 1 }
        },
        {
            name  => 'customfield7',
            value => { value => $customfields{7}, cdata => 1 }
        },
        {
            name  => 'customfield8',
            value => { value => $customfields{8}, cdata => 1 }
        },
        {
            name  => 'customfield9',
            value => { value => $customfields{9}, cdata => 1 }
        },
        {
            name  => 'customfield10',
            value => { value => $customfields{10}, cdata => 1 }
        }
    );

    # Build XML from structure
    for my $field (@fields) {
        my $tag   = $xml->createElement( $field->{name} );
        my $value = $field->{value};
        if ( defined( $value->{'value'} ) && $value->{'value'} ne '' ) {
            if ( $value->{'cdata'} ) {
                my $data = XML::LibXML::CDATASection->new("$value->{'value'}");
                $tag->appendChild($data);
            }
            else {
                $tag->appendTextNode( $value->{'value'} );
            }
        }
        $root->appendChild($tag);
    }

    # Retrieve default field values
    my $DefaultVATDesc = $self->retrieve_data('DefaultVATDesc');
    my $DefaultVATCode = $self->retrieve_data('DefaultVATCode');
    my $DefaultVATRate = $self->retrieve_data('DefaultVATRate');

    # Add the accountlines to pay off
    my @accountline_ids = $cgi->multi_param('accountline');
    $debug
      and warn "Adding accountlines to transaction: "
      . join( ', ', @accountline_ids );
    my $accountlines = $schema->resultset('Accountline')
      ->search( { accountlines_id => \@accountline_ids } );
    my $now               = DateTime->now;
    my $dateoftransaction = $now->ymd('-') . ' ' . $now->hms(':');

    my $sum = 0;
    for my $accountline ( $accountlines->all ) {

        # Track sum
        my $amount = sprintf "%.2f", $accountline->amountoutstanding;
        $sum = $sum + $amount;

        # Build payments block
        ######################
        my $payments = $xml->createElement('payments');
        $payments->setAttribute( 'id'   => $accountline->accountlines_id );
        $payments->setAttribute( 'type' => 'PN' );

        my $debit_type_code =
            $self->_version_check('19.11.00')
          ? $accountline->debit_type_code->code
          : $accountline->accounttype;
        if ( exists( $system_type_map{$debit_type_code} ) ) {
            $debit_type_code = $system_type_map{$debit_type_code};
        }
        $payments->setAttribute( 'payoption' => $debit_type_code );

        my $description = $xml->createElement("description");
        if ( defined( $accountline->description )
            && $accountline->description ne '' )
        {
            my $data_description = $accountline->description;
            my $data =
              XML::LibXML::CDATASection->new(
                encode_entities($data_description) );
            $description->appendChild($data);
        }
        $payments->appendChild($description);

        # Build payment block
        my $payment = $xml->createElement("payment");
        $payment->setAttribute( 'payid' => $accountline->accountlines_id );

        my $custom1 = $self->retrieve_data('payment_customfield1');
        if ($custom1) {
            my $customfield1 = $xml->createElement("customfield1");
            $customfield1->appendTextNode($custom1);
            $payment->appendChild($customfield1);
        }

        my $amounttopay = $xml->createElement("amounttopay");
        $amounttopay->appendTextNode($amount);
        $payment->appendChild($amounttopay);

        my $amounttopayvat = $xml->createElement("amounttopayvat");
        $amounttopayvat->appendTextNode('0');
        $payment->appendChild($amounttopayvat);

        my $amounttopayexvat = $xml->createElement("amounttopayexvat");
        $amounttopayexvat->appendTextNode($amount);
        $payment->appendChild($amounttopayexvat);

        my $vatdesc = $xml->createElement("vatdesc");
        $vatdesc->appendTextNode($DefaultVATDesc);
        $payment->appendChild($vatdesc);

        my $vatcode = $xml->createElement("vatcode");
        $vatcode->appendTextNode($DefaultVATCode);
        $payment->appendChild($vatcode);

        my $vatrate = $xml->createElement("vatrate");
        $vatrate->appendTextNode($DefaultVATRate);
        $payment->appendChild($vatrate);

        my $dateofpayment = $xml->createElement("dateofpayment");
        $dateofpayment->appendTextNode($dateoftransaction);
        $payment->appendChild($dateofpayment);

        my $editable = $xml->createElement("editable");
        $editable->setAttribute( 'minamount' => "0" );
        $editable->setAttribute( 'maxamount' => "0" );
        $editable->appendTextNode(0);
        $payment->appendChild($editable);

        my $mandatory = $xml->createElement("mandatory");
        $mandatory->appendTextNode(1);
        $payment->appendChild($mandatory);

        # Add 'payment' to 'payments' block
        $payments->appendChild($payment);

        # Add 'payments' to 'root' block
        $root->appendChild($payments);
    }

    # Add signature to xml
    $sum = sprintf "%.2f", $sum;
    $debug and warn "Total to pay" . $sum;
    my $msgid =
      md5_hex( $self->retrieve_data('WPMClientID')
          . $transaction_id
          . $sum
          . $self->retrieve_data('WPMSecret') );
    $root->setAttribute( 'msgid' => "$msgid" );

    # Finalise XML Document
    $xml->setDocumentElement($root);
    my $string = $xml->toString(2);

    $template->param(
        WPMPathway => $self->retrieve_data('WPMPathway'),
        XMLPost    => $string
    );

    print $cgi->header();
    print $template->output();
}

=head2 opac_online_payment_end

  Complete online payment process

=cut

sub opac_online_payment_end {
    my ( $self, $args ) = @_;

    $debug and warn "Inside opac_online_payment_end for: " . caller . "\n";
    my $cgi = $self->{'cgi'};

    my ( $template, $borrowernumber ) = get_template_and_user(
        {
            template_name   => $self->mbf_path('opac_online_payment_end.tt'),
            query           => $cgi,
            type            => 'opac',
            authnotrequired => 0,
            is_plugin       => 1,
        }
    );

    my $transaction_id = $cgi->param('transaction_id');

    # Check payment went through here
    my $table = $self->get_qualified_table_name('wpm_transactions');
    my $dbh   = C4::Context->dbh;
    my $sth   = $dbh->prepare(
        "SELECT accountline_id FROM $table WHERE transaction_id = ?");
    $sth->execute($transaction_id);
    my ($accountline_id) = $sth->fetchrow_array();

    my $line =
      Koha::Account::Lines->find( { accountlines_id => $accountline_id } );
    my $transaction_value  = $line->amount;
    my $transaction_amount = sprintf "%.2f", $transaction_value;
    $transaction_amount =~ s/^-//g;

    if ( defined($transaction_value) ) {
        $template->param(
            borrower      => scalar Koha::Patrons->find($borrowernumber),
            message       => 'valid_payment',
            message_value => $transaction_amount
        );
    }
    else {
        $template->param(
            borrower => scalar Koha::Patrons->find($borrowernumber),
            message  => 'no_amount'
        );
    }

    print $cgi->header();
    print $template->output();
}

## If your plugin needs to add some javascript in the OPAC, you'll want
## to return that javascript here. Don't forget to wrap your javascript in
## <script> tags. By not adding them automatically for you, you'll have a
## chance to include other javascript files if necessary.
sub opac_js {
    my ($self) = @_;

    # We could add in a preference driven 'enforced pay all' option here.
    return q|
        <script></script>
    |;
}

## If your tool is complicated enough to needs it's own setting/configuration
## you will want to add a 'configure' method to your plugin like so.
## Here I am throwing all the logic into the 'configure' method, but it could
## be split up like the 'report' method is.
sub configure {
    my ( $self, $args ) = @_;
    my $cgi = $self->{'cgi'};

    unless ( $cgi->param('save') ) {
        my $template = $self->get_template( { file => 'configure.tt' } );

        ## Grab the values we already have for our settings, if any exist
        $template->param(
            enable_opac_payments =>
              $self->retrieve_data('enable_opac_payments'),
            WPMClientID          => $self->retrieve_data('WPMClientID'),
            WPMSecret            => $self->retrieve_data('WPMSecret'),
            WPMPathway           => $self->retrieve_data('WPMPathway'),
            WPMPathwayID         => $self->retrieve_data('WPMPathwayID'),
            WPMDepartmentID      => $self->retrieve_data('WPMDepartmentID'),
            DefaultVATDesc       => $self->retrieve_data('DefaultVATDesc'),
            DefaultVATCode       => $self->retrieve_data('DefaultVATCode'),
            DefaultVATRate       => $self->retrieve_data('DefaultVATRate'),
            customfield1         => $self->retrieve_data('customfield1'),
            customfield2         => $self->retrieve_data('customfield2'),
            customfield3         => $self->retrieve_data('customfield3'),
            customfield4         => $self->retrieve_data('customfield4'),
            customfield5         => $self->retrieve_data('customfield5'),
            customfield6         => $self->retrieve_data('customfield6'),
            customfield7         => $self->retrieve_data('customfield7'),
            customfield8         => $self->retrieve_data('customfield8'),
            customfield9         => $self->retrieve_data('customfield9'),
            customfield10        => $self->retrieve_data('customfield10'),
            payment_customfield1 =>
              $self->retrieve_data('payment_customfield1'),
        );

        print $cgi->header();
        print $template->output();
    }
    else {
        $self->store_data(
            {
                enable_opac_payments =>
                  scalar $cgi->param('enable_opac_payments'),
                WPMClientID          => scalar $cgi->param('WPMClientID'),
                WPMSecret            => scalar $cgi->param('WPMSecret'),
                WPMPathway           => scalar $cgi->param('WPMPathway'),
                WPMPathwayID         => scalar $cgi->param('WPMPathwayID'),
                WPMDepartmentID      => scalar $cgi->param('WPMDepartmentID'),
                DefaultVATDesc       => scalar $cgi->param('DefaultVATDesc'),
                DefaultVATCode       => scalar $cgi->param('DefaultVATCode'),
                DefaultVATRate       => scalar $cgi->param('DefaultVATRate'),
                customfield1         => scalar $cgi->param('customfield1'),
                customfield2         => scalar $cgi->param('customfield2'),
                customfield3         => scalar $cgi->param('customfield3'),
                customfield4         => scalar $cgi->param('customfield4'),
                customfield5         => scalar $cgi->param('customfield5'),
                customfield6         => scalar $cgi->param('customfield6'),
                customfield7         => scalar $cgi->param('customfield7'),
                customfield8         => scalar $cgi->param('customfield8'),
                customfield9         => scalar $cgi->param('customfield9'),
                customfield10        => scalar $cgi->param('customfield10'),
                payment_customfield1 =>
                  scalar $cgi->param('payment_customfield1'),
                last_configured_by => C4::Context->userenv->{'number'},
            }
        );
        $self->go_home();
    }
}

## This is the 'install' method. Any database tables or other setup that should
## be done when the plugin if first installed should be executed in this method.
## The installation method should always return true if the installation succeeded
## or false if it failed.
sub install() {
    my ( $self, $args ) = @_;

    $self->store_data(
        {
            DefaultVATDesc => 'Exempt',
            DefaultVATCode => 'E',
            DefaultVATRate => 0,
        }
    );

    my $table = $self->get_qualified_table_name('wpm_transactions');

    return C4::Context->dbh->do( "
        CREATE TABLE IF NOT EXISTS $table (
            `transaction_id` INT( 11 ) NOT NULL AUTO_INCREMENT,
            `accountline_id` INT( 11 ),
            `updated` TIMESTAMP,
            PRIMARY KEY (`transaction_id`)
        ) ENGINE = INNODB;
    " );
}

## This is the 'upgrade' method. It will be triggered when a newer version of a
## plugin is installed over an existing older version of a plugin
#sub upgrade {
#    my ( $self, $args ) = @_;
#
#    my $dt = dt_from_string();
#    $self->store_data(
#        { last_upgraded => $dt->ymd('-') . ' ' . $dt->hms(':') } );
#
#    return 1;
#}

## This method will be run just before the plugin files are deleted
## when a plugin is uninstalled. It is good practice to clean up
## after ourselves!
#sub uninstall() {
#    my ( $self, $args ) = @_;
#
#    my $table = $self->get_qualified_table_name('mytable');
#
#    return C4::Context->dbh->do("DROP TABLE $table");
#}

1;
